### DISCUSSION TITLE
Revisiting the Usage of Block Variations

### DISCUSSION THREAD

**Discussion by nerrad** [woocommerce/woocommerce#44969] [WooCommerce, Blocks, and Site Editing]
In #7954, the team is looking into removing margins from CSS with blocks to better support Global Styles. While there were a number of approaches discussed, the larger issue that surfaced as I was thinking through and discussing approaches with the team are a number of limitations we're running into with how we *want* to use Block Variations vs how they are *intended* to be used.

**Before we dive too much further into our use of this API, I think we need to take a step back and consider the intended use (and ensure that's a correct understanding), and validate how that impacts our current and planned needs for the blocks we have using this API.**

To make it extra clear: as an outcome from the discussion generated from this post, what I'm aiming for:

- Facilitating additional clarity/insights from folks working on the GB project about what has been outlined here and ensure we have a correct understanding of the Block Variations API.
- With this correct understanding, have a better idea of where we say yes to using the Block Variations API and where we say no.
- Identify some next steps around our *existing* Block Variation usage and where we switch to standalone blocks.
- Identify how we can motivate/contribute improvements to either the Block Variation API *or* specific Blocks to support more useful extensibility for the needs of WordPress plugin developers in a way that *enhances* the interfaces for users of the extended blocks.

> **Note** To be extra clear, this is not referring to block *style* variations. What I'm referring to here as Block Variations are blocks that are registered as a variation of another block. For example, the Products Block is a variation of the WP Query Loop Block. The Product Title block is a variation of the Post Title block etc.

For reference, these are the current WooCommerce Blocks that are variations of a WP Core Block (in trunk as of this post):

| WooCommerce Block Variation | WP Core Block |
|-------------|---------|
| Product Summary | Post Excerpt |
| Product Template | Post Template |
| Product Title | Post Title |
| Product Query | Query Loop |
| Related Products | Query Loop |

## Principles of Block Variations informing the APIs Intended Usage

From the [official docs](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-variations):

> Block Variations is the API that allows a block to have similar versions of it, but all these versions share some common functionality. Each block variation is differentiated from the others by setting some initial attributes or inner blocks.

An example in WP core of the variation usage are the embed blocks where they all share the same common boilerplate for getting embeds but the individual implementations differ on the provider (via the `providerNameSlug` attribute). Using this description, examples from implementations in WP, various conversations with project maintainers, and the Woo team's learnings and discussions from using them, there have emerged a set of principles around the intended use of Block Variations.

### 1. Use Block Variations when you need to create multiple instances of a block with the same functionality but which have small differences in appearance, or attributes.

For the most part, Block Variations should not require any changes to the block toolbar or the Block Inspector panel to achieve these small differences and the degree of difference applies to the logic inherent in the variation Inner Blocks or embedded behaviour.

The benefit achieved through Block Variations here is *reusability* and *maintenance*. With reusability, there is less code, and with maintainability, any updates to the base block will automatically apply to all its variations.

### 2. Use Block Variations to ensure a consistent *default* user experience across different implementations of the base block and it's variants.

This allows for consistent default design and behaviour defined at the base block level and across all variants. This is currently enforced by:

- `theme.json` is not able to style individual Block Variations.
- Block Variations are not configurable via the Global Styles interface.

User level (or Pattern level) customization is still possible for Block Variations but limited to:

- the controls provided by the base block.
- any "block supports" or restrictions implemented by the theme for the base block.

### 3. Block Variations do not implement an inheritance model.

Block variations do not *extend* the main block. Block variations never add new functionality.

> Note: This is a natural outcome of the previous two principles but I think is something that is not clear in the Block Variation docs so worth calling out explicitly.

## Implications

The intent behind the Block Variation API is one that I don't think has been fully clear to the team and naturally as a result we approached treating it as an inheritance model. One in which we'd identify the gaps for things not extensible and work within GB core to address those gaps.

Initially, the gaps we exposed existed primarily with our usage of the Query Loop block and resulted in a number of changes including:

- Implementing a filter on the server generated query paramaters ([GB PR](https://github.com/WordPress/gutenberg/pull/43590/))
- Implementing an `allowedControls` API surface for Query Loop variations to indicate what controls are available with the variation ([GB PR](https://github.com/WordPress/gutenberg/pull/43632))
- The introduction of an API for passing along extra query args in REST API calls (which allow for editor previews matching what is rendered on the frontend). ([GB PR](https://github.com/WordPress/gutenberg/pull/44093))
- Suggesting active variation patterns for the variation ([GB PR](https://github.com/WordPress/gutenberg/pull/44197))
- The addition of docs to the GB project with instructions on how to customize and extend the Query Loop block ([GB PR](https://github.com/WordPress/gutenberg/pull/44137))
- Improve the labeling of a Query Loop variation when it is picked via the Block Variation picker ([GB PR](https://github.com/WordPress/gutenberg/issues/46367))
- Ensure children of block variations refer to parents with original name ([GB PR](https://github.com/WordPress/gutenberg/issues/46833))

**Proposed and not implemented:**

- Enabling full customization of Query Loop block inspector controls ([GB PR](https://github.com/WordPress/gutenberg/pull/43675))
- Try Query Loop filters slot ([GB PR](https://github.com/WordPress/gutenberg/pull/43684))
- Block variations are missing separate CSS selectors ([Issue](https://github.com/woocommerce/woocommerce-blocks/issues/8411))
- Block variations are missing Global Styles support ([Issue](https://github.com/WordPress/gutenberg/issues/47324))
- Add support for the `ancestor` and `parent` settings for Block Variations ([Issue](https://github.com/WordPress/gutenberg/issues/48424))

Reviewing the above PRs and issues indicate some common patterns:

- There are limitations with some complex blocks that impact the the block output for consumers of the relevant API. For instance, some gaps exist for the utilization of custom post type support for Query Loop implementations. While Block Variations are a way to overcome some of these gaps, arguably the extensibility of the main block needs improvements and the *intent* of the Block Variation API isn't matching the way we've been using it.
- Some existing gaps are *unlikely* to be addressed given the intent of the Block Variation API. Particular examples are, slots for Query Loop, Global Styles support.

As a whole then, there are some usability implications for our current usage of Block Variations:

- Block variations *are* available in the inserter individually, but there's no clear indicator that they are a "special" type of block, and thus not being able to see them in places like the Global Styles interface (or Style Book) could be perceived as a bug.
- Making changes to the main block styles could lead to unexpected behaviour for the variant blocks. In what contexts is this okay vs what contexts will this be perceived as a bug by users? An example might be changing the global styles for the "Post Title" block and those changes appearing on every variation of that block (which currently includes in Woo's case, "Product Title").
- Different user interfaces that are added for blocks could have unexpected implications for any variants. For example, the recently added "Apply Globally" button for pushing individual block style changes to Global Styles. Applying changes this way will apply to *all* variants of the main block.
- The Products Query block has multiple panels for filtering functionality because currently new filtering options cannot be added to the existing filter panel.
- Some variations display placeholder content that is contextual to the parent block but not the variant (and is not editable by the variant). For instance the placeholder content for the Product Title block (a variant of the Post Title block) is "Post Title", or the placeholder content for the Product Summary block (a variant for the Post Excerpt block) is "This is the Post Excerpt block, it will display the excerpt from single posts. If there are any Custom Post Types with support for excerpts, the Post Excerpt block can display excerpts of those entries as well". Some of these terms (eg "Custom Post Types") are meaningless to a new WordPress user (let alone merchants that came to Woo first, not WP first). GB could potentially improve this by making the "Post Type" type blocks more contextually aware and use registered labels for the post type, but that is only one part of the problem.

## Next steps?

**A loose compendium of where we should go from here:**

- If the principles I've captured here about block variations are useful, this might be a good update to its documentation.
- @shaunandrews, I think the global styles support for block variations is a significant factor here. I'm leaning on you a bit for how much of a factor this is and which blocks we should revert to be standalone and which blocks should stay as is given the limitations I've outlined in this post.
- There is so much value inherent in the logic provided out of the box for the Query Loop block that I'm hesitant to outright decide the Products Block should not be a variant of it (especially given the work that has already been done to make it more usable). IMO it's a special case so I've extracted thoughts about it into a separate section below.
- Where we end up *not* using the Block Variation API, we should definitely look at what we can expose as components/shared functionality from the Core Blocks so much of it can be repackaged in new blocks (especially some of the more complex features).
- Post Type type blocks would ideally infer a lot more from the registered post type details (labels etc). This could be a potential path for either automatically exposing variations in the Block editor or new "types" of blocks specific to Query related data. The API related to this could be much more targeted (and mostly automated from the registered data). This likely would require larger effort within Gutenberg to implement though so in the short term, it's not a solution for Woo.

#### The special case of the Products Block

The usability of the inspector controls and the ability to extend with additional filters is a pretty significant blocker at the moment (one I don't think is Woo specific). Our options are (without any changes in GB):

- Limit the functionality to users for the Products block.
- Add additional functionality via existing extensibility options (additional panel for filters) and accept the less than ideal interface exposed to users
- Build a standalone block that potentially has to duplicate a lot of the functionality in the Query Loop with the tradeoff of complete customizability of the user interface. 

On the surface, its looking like the third option will be the best approach here.

- The block would be very targeted (no need to support multiple post types like the existing block). This limits some of the code duplication that would occur.
- On the surface, it looks like we could tap into the server side rendering/extensibility (although that could depend on how much of the next point we utilize).
- We can more easily solve the multiple REST request problem in the editor by utilizing the `wc/v3/products` endpoint for the entity retrieval. Which would also have the side-effect of better supporting a potential "universal block" idea for blocks that are context aware of what editor they are in. The in progress block powered product editor is using `wc/products`.
- Global Styles support.

Of course there still would need to be decisions on:

- Which core blocks can we still use as inner blocks (eg. Pagination)? I think we should as much as possible.
- Do we follow the same structure for inner blocks (eg. Query Loop -> Post Template -> Inner Blocks)? I think yes.

And downsides:

- We won't automatically inherit improvements to the Query Loop block.
- There could be some duplication of code (although we could *likely* use the same endpoints for the query handling server side).
- Implementing a migration/transformation path for existing uses of block variations to standalone blocks is additional work with potential risks. Still doable I think.

## Feedback!

- @mtias, @richtabor or others (@gziolo, @ntsekouras) will you provide some feedback on what I've written about the intent of block variations? Is this accurate? Are there things I've identified that might not be accomplished by block variations but still can be resolved some way in GB via the Block API?
- @woocommerce/woo-fse is there anything I've missed in this discussion that you can expand on?
- Folks in the larger ecosystem who might land on this thread, what have you encountered with the Block Variation API that might provide additional insight into understanding their intent?

## References

- [Moving Core block styling to JSON](https://make.wordpress.org/core/2022/08/01/moving-core-block-styling-to-json/)
- Slack Convo (internal): p1680523594810199-slack-C02UBB1EP - This convo was mostly just around #7954 and the potential options to take there.

**Comment by chrisdesrochers** [comment]
Great summary @nerrad! Gaining clarity here will be critical to our continued use of the Block Variations API and will help us scope any remediation needed to ensure our blocks are compliant going forward.

**Reply by gziolo** [reply]
I wanted to echo the comment as I strongly agree that @nerrad did an outstanding job collecting all the information in one place about block variations. We definitely need to figure out how flexible the API should be and document it better to give proper guidance to developers trying to take advantage of it.

With that said, I need to clearly state that the Products List variation is probably one of the most complex one for the Query Loop block😅

**Comment by danielwrobert** [comment]
> @woocommerce/woo-fse is there anything I've missed in this discussion that you can expand on?

Nothing comes to mind for me – you’ve covered about everything I can think of and I agree with all of your points and potential next steps. Nice work!

> • Which core blocks can we still use as inner blocks (eg. Pagination)? I think we should as much as possible.
> • Do we follow the same structure for inner blocks (eg. Query Loop -> Post Template -> Inner Blocks)? I think yes.

I agree with both of these. If we determine it’s best to take a more compositional approach, I think it would still be best to use as much from core as we can. Many core blocks are general enough that they still make contextual sense for us to use as inner blocks.

> • Implementing a migration/transformation path for existing uses of block variations to standalone blocks is additional work with potential risks. Still doable I think.

Agreed. Additionally, I think the sooner we decide if this is our path forward, the easier it will be to course correct on the work that we’ve already put into the Products Block. Sure, it’s not ideal and we’ve already invested a lot of time with the Block Variation API route. But if this is the better way to go, as opposed to wrestling with the Block Variation API because we’re trying to use it in a way that it was not intended to be used, we’ll be much better off in the long run.

**Comment by ntsekouras** [comment]
Thanks for the great write-up @nerrad!

I agree with your points regarding block variation API with a small note about `the inheritance model`. In my mind it was never the case, but taking into account all your work with the API, any improvements/clarifications in the docs are most welcome.

The second point about block variations support in Global Styles I think it's an important one and very valuable. I don't know all the nuances about such implementation, but is something that we should explore soon. I'll cc @oandregal for first thoughts about it.

I don't think I have much more to say, as you have described both pros and cons very well, wether keeping variations or not. While it's very powerful that you cat get things for 'free' from core, development in such a big open source project is sometimes slower though.

You'll know best if you make a standalone block for `Products`. In general I think there is no frantic development in `Query Loop` right now, as it supports quite a few things already(that means that copying some code is a viable scenario) and `Products` has specific needs that would need a lot of time to implement in core and when they land, they might not be exactly what Woo would need. For example the [ancestor/parent variation issue](https://github.com/WordPress/gutenberg/issues/48424) is more complex than it sounds, as it essentially also involves the relationship of variations that could refer to other variations(Product Title-> Products Loop).

Having said that the block variations API is evolving and will become more powerful in time(an example would be this [new PR](https://github.com/WordPress/gutenberg/pull/49728) for block styles for specific block variations). Woo drove a lot of useful extensibility enhancements and could drive even more, to speed things up.

**Reply by nerrad** [reply]
Your help has been really instrumental in getting things this far Nik, thanks! I'm definitely supportive of helping make improvements to the API around block variations where we can. But to a degree, some of the improvements shipped, and some of the ones that are planned (block styles for specific block variations) seem to contradict the overall intent of Block variations? The time for getting enhancements in isn't as much of an issue for us, however we don't want to drive enhancements that are not congruent with the overall intent of the API. 

Which is pretty much the crux of this post - it appears the things Woo _needs_ Block Variations to allow for as an API, won't be things supported in the Block Variations API because it contradicts the intent of the API.

To clarify one point:

> I agree with your points regarding block variation API with a small note about the inheritance model. In my mind it was never the case, but taking into account all your work with the API, any improvements/clarifications in the docs are most welcome.

Do you mean that in your mind it was never the case that the Block Variation was an inheritance model? If so, yeah I can see that that docs don't explicitly suggest that. However, they don't explicitly deny it either and I think it's relatively easy for folks initially not familiar with the API to infer this as the model as they begin building with it.

**Reply by ntsekouras** [reply]
>seem to contradict the overall intent of Block variations

Why do think that? Block styles, Global Styles for variations are 'appearance' differences. Also having a consistent default user experience can work well, but if you also want to `extend` the functionality of a block, it's okay to diverge from some UI settings(hide/show etc..). It depends on the complexity of each use case and many cases that might seem at first as block variation API related, could be just `Blocks API` issues. An example would be the conditional showing/hiding controls that we implemented specifically for Query for now, and this is something that all blocks should have.

>Do you mean that in your mind it was never the case that the Block Variation was an inheritance model?

That's a tricky one. Yes and no. I mean, the way to `inherit` and add new things is through the extension of blocks, where there are many ways to do so. Variations can add to that by providing other useful affordances, like of course the reusability of code, easier insertion/transformation flows, etc..

For complex use cases, all of that is needed, in combination with the extension of blocks, to try to get as much as you can with both APIs(extend+variations), and that what Woo does in `Products`. In such cases the blocks could be 'treated' as autonomous blocks, even though they are not entirely. What we are missing right now, is some more granular control, like the Global Styles case or be able to disable specific block variations, etc..

I hope I'm not rumbling here, as I try to be helpful 😄

**Reply by nerrad** [reply]
> I hope I'm not rumbling here, as I try to be helpful 😄

Nope that's helpful. 

> Why do think that? Block styles, Global Styles for variations are 'appearance' differences.

FWIW I agree with you, but another perspective is that global style support for variations relaxes the rules around consistent defaults for all variations of the main block. I'm happy if the lack of support for theme.json and Global Styles for Variations is an oversight and not intentional. If so, then the second principle I mentioned in my post is not correct (as an intent for the API).

> I mean, the way to inherit and add new things is through the extension of blocks, where there are many ways to do so. Variations can add to that by providing other useful affordances, like of course the reusability of code, easier insertion/transformation flows, etc..

Right, this is where things get tricky. What exactly is meant by "Block variations do not implement an inheritance model" (my paraphrasing of something @mtias explicitly mentioned in a conversation)? This could be interpreted various ways:

- Block variations _don't_ inherit from the main block. 
- Block variations _don't_ extend the main block.
- Block variations _don't_ add new functionality.

All of those have different implications for the consumers of the API and expectations around what would be acceptable contributions to enhancing the API.

On the surface, it also doesn't appear everyone contributing to the Block Variations API is on the same page which adds to the complexity for consumers implementing it.

**Reply by gziolo** [reply]
> For example the https://github.com/WordPress/gutenberg/issues/48424 is more complex than it sounds, as it essentially also involves the relationship of variations that could refer to other variations(Product Title-> Products Loop).

I agree that it’s something that might have some unexpected consequences as folks might not only want to reference other variations but they also could list other blocks that might not necessarily work when potentially transforming the variation to something else. I’m talking here about other blocks that might use different setup.

**Comment by imanish003** [comment]
Thank you for initiating this important discussion on the usage of Block Variations. Your comprehensive analysis and suggestions provide a solid foundation for the discussion, and I look forward to seeing the outcomes and decisions resulting from this conversation 🙌🏻

**Comment by nefeline** [comment]
This is a fantastic summary of block variations, and I have also learned a few things by reading through this discussion: thank you! 

> https://github.com/orgs/woocommerce/teams/woo-fse is there anything I've missed in this discussion that you can expand on?

I can't think of anything missing, but I'd like to share my 2 cents:

We recently worked on the [Single Product block](https://github.com/woocommerce/woocommerce-blocks/pull/8610), which relies on some block variations such as the Product Summary (Post Excerpt) and the Product Title (Post title). For those blocks which are more atomic and an exact match of what the core of WordPress renders (but in the context of a single product (which is just a custom post type) rather than a single post), I'm more inclined to stick with variations as there's no need to reinvent the wheel: especially with @ntsekouras ' upcoming work https://github.com/WordPress/gutenberg/pull/49728, adding support for block variation styles. 

In such cases, though, I agree that allowing for some fundamental changes, such as modifying the placeholder content (as you flagged in the discussion), is essential.

I didn't have the opportunity to work closely with the Products Block just yet, but based on what I saw during code reviews and by reading/participating in our team's discussions and struggles along the way (as very well documented here), I think it's fair for us to deviate from the Query Loop and have our dedicated codebase/implementation instead. Products are the heart of our product (Woo!), and IMO, it would be great to have more autonomy over how we want them to be displayed and customized, which is, in many cases, not fully aligned with what a regular/native blog loop expects/needs for listing regular posts. In other words, I agree with:

> Build a standalone block that potentially has to duplicate a lot of the functionality in the Query Loop with the tradeoff of complete customizability of the user interface.

**Reply by gziolo** [reply]
> In such cases, though, I agree that allowing for some fundamental changes, such as modifying the placeholder content (as you flagged in the discussion), is essential.

The issue with wording in placeholders sounds like a bug, as even in WP core there is a distinction between posts and pages.

Products list block, the wrapping one, turns out to be way more complex than the Query Loop block shipped with WP core, so maybe we won’t ever be able to offer an extensibility model that satisfies all the requirements that are covered in the description. It’s definitely easier to achieve for more granular blocks used.

**Comment by oandregal** [comment]
> theme.json is not able to style individual Block Variations.
> Block Variations are not configurable via the Global Styles interface.

Global styles can be thought of as a sophisticated way to abstract CSS selectors. Blocks or elements are high-level selectors, so to speak. As long as a piece of HTML can be addressable via CSS as a standalone unit, it can work in any global styles device (`theme.json`, the global styles sidebar, etc.).

By what I've seen, it seems to me that variations only exist at "insertion time". Once the block is inserted, the content is unaware of them: all variations share the same CSS selector. For example, if I add a "query block" and a "posts list" block in the same page, how can I style them differently?

So, the relationship between global styles & block variations boils down to whether block variations can be addressable via a unique CSS selector. Is that something we have/want/plan to? If that's something in the roadmap, we should look at the recently stabilized API for [blocks to declare its CSS selectors](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-selectors/).

**Reply by ntsekouras** [reply]
Thanks for the input @oandregal! 

>Is that something we have/want/plan to?

No idea yet, but it's nice to see the technical nuances about this. So, if we had a specific css class in the markup of the block. Would that be possible and easy to add support? Like if that block has also that class, prioritize those styles.

I had explored adding a css class dynamically based on an active block variation, and while in JS is possible, we have an issue in PHP because the `isActive` property can be either an <string[]> or a function. In the cases of being a JS function, we cannot do it PHP.

**Reply by gziolo** [reply]
There is this idea tracked in the [Block API issue](https://github.com/WordPress/gutenberg/issues/41236):

> Add handling for block type name aliases used with block variations (https://github.com/WordPress/gutenberg/issues/43743). They should fall back if necessary to the original block type name, e.g. `woocommerce/product-list` that can get resolved to the `product-list` variation added to the `core/query` block.

If we had something like that in place, then it would automatically generate a unique class name for the variation.

**Reply by oandregal** [reply]
Right. The data used by global styles is the selector & supports of a given block type, which is queried from the block registry. As long as there is a way for the server and client to have access to selector & supports for a given block variation, it can be exposed in any device (theme.json, sidebar).

**Reply by mtias** [reply]
Style variations have a class (they were, essentially, as introduced, nothing but a "class" abstraction). Improvements to style variation engine (assimilating it over block variations in general) — see #49602 — can result in the specific class hook existing for global styles controls.

**Comment by nerrad** [comment]
Thanks, everyone for participating in this discussion! Some decisions have come out of this that I wanted to follow up on:

- The Products Block will be shifted to be a standalone block that will still utilize similar functionality (and components where possible). This will allow for much more flexibility that won't be available via the Query Loop block because of the direction for Block variations (particularly in the area of the inspector controls). 
- For now, we'll keep using variations for things like Post Title (Product Title) and Post Excerpt (Product Summary) while working to address the existing issues around placeholder content etc upstream in the GB repository. Those improvements would ideally be available for all custom post types so there are some opportunities to improve there.
- We're also continuing to monitor the support for Block variation global style support and help test where we can. This will be relevant for our continued use of Block Variations for some of the post-type related blocks.
